src/Simulator/IncrementalRedstoneSimulator.cpp

```diff
@@ -77,68 +77,60 @@ void cIncrementalRedstoneSimulator::RedstoneAddBlock(int a_BlockX, int a_BlockY,
-		}
-	), RepeatersDelayList.end());
-
-	if (a_OtherChunk != nullptr)
+	if (a_NeighborChunk != nullptr)
 	{
 		// DO NOT touch our chunk's data structure if we are being called with coordinates from another chunk - this one caused me massive grief :P
 		return;
 	}
 
-	cCoordWithBlockAndBoolVector & RedstoneSimulatorChunkData = ((cIncrementalRedstoneSimulator::cIncrementalRedstoneSimulatorChunkData *)a_Chunk->GetRedstoneSimulatorData())->m_ChunkData;
-	for (auto & itr : RedstoneSimulatorChunkData)
+	auto & SimulatedPlayerToggleableBlocks = static_cast<cIncrementalRedstoneSimulatorChunkData *>(a_OriginalChunk->GetRedstoneSimulatorData())->m_SimulatedPlayerToggleableBlocks;
+	if (DoesIgnorePlayerToggle(Block))
 	{
-		if ((itr.x == RelX) && (itr.y == a_BlockY) && (itr.z == RelZ))  // We are at an entry matching the current (changed) block
-		{
-			if (!IsAllowedBlock(Block))
-			{
-				itr.DataTwo = true;  // The new blocktype is not redstone; it must be queued to be removed from this list
-			}
-			else
-			{
-				itr.DataTwo = false;
-				itr.Data = Block;  // Update block information
-			}
-			return;
-		}
+		// Initialise the toggleable blocks list so that trapdoors etc. aren't reset on restart (#1887)
+		SimulatedPlayerToggleableBlocks.emplace(
+			a_RelBlockPosition,
+			AreCoordsDirectlyPowered(a_RelBlockPosition.x, a_RelBlockPosition.y, a_RelBlockPosition.z, a_OriginalChunk) || AreCoordsLinkedPowered(a_RelBlockPosition.x, a_RelBlockPosition.y, a_RelBlockPosition.z, a_OriginalChunk)
+			);  // This map won't insert if key already present, so no need to check
+	}
+	else
+	{
+		SimulatedPlayerToggleableBlocks.erase(a_RelBlockPosition);
+	}
+
+	if ((Block != E_BLOCK_REDSTONE_REPEATER_ON) && (Block != E_BLOCK_REDSTONE_REPEATER_OFF))
+	{
+		static_cast<cIncrementalRedstoneSimulatorChunkData *>(a_OriginalChunk->GetRedstoneSimulatorData())->m_RepeatersDelayList.erase(a_RelBlockPosition);
 	}
 
+	auto & RedstoneSimulatorChunkData = static_cast<cIncrementalRedstoneSimulatorChunkData *>(a_OriginalChunk->GetRedstoneSimulatorData())->m_ChunkData;
+	auto Iterator = RedstoneSimulatorChunkData.find(a_RelBlockPosition);
 	if (!IsAllowedBlock(Block))
 	{
+		if (Iterator != RedstoneSimulatorChunkData.end())
+		{
+			Iterator->second.second = true;  // The new blocktype is not redstone; it must be queued to be removed from this list
+		}
 		return;
 	}
-
-	cCoordWithBlockAndBoolVector & QueuedData = ((cIncrementalRedstoneSimulator::cIncrementalRedstoneSimulatorChunkData *)a_Chunk->GetRedstoneSimulatorData())->m_QueuedChunkData;
-	for (const auto & itr : QueuedData)
+	else
 	{
-		if ((itr.x == RelX) && (itr.y == a_BlockY) && (itr.z == RelZ))
+		if (Iterator != RedstoneSimulatorChunkData.end())
 		{
-			// Can't have duplicates in here either, in case something adds the block again before the structure can written to the main chunk data
+			Iterator->second.second = false;
+			Iterator->second.first = Block;  // Update block information
```

Comment:
No comments do not. And yes having 10 additional files is superior to the current situation, because when working out whats going on, you only have to look at a subset of the code.